### PR TITLE
REF: Move generic methods to aggregation.py

### PR DESCRIPTION
**Pull request by charlesdong1991**
xref #29116

https://github.com/pandas-dev/pandas/pull/29116#pullrequestreview-340272465

based on @jreback comment, this PR is a precursor PR for #29116 to make PR smaller and more readable

### CODE REVIEW

File: pandas/core/aggregation.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    --------
+    >>> normalize_keyword_aggregation({'output': ('input', 'sum')})
+    ({'input': ['sum']}, ('output',), [('input', 'sum')])
+    """
+    # Normalize the aggregation functions as Mapping[column, List[func]],
+    # process normally, then fixup the names.
+    # TODO: aggspec type: typing.Dict[str, List[AggScalar]]
+    # May be hitting https://github.com/python/mypy/issues/5958
+    # saying it doesn't have an attribute __name__
+    aggspec: DefaultDict = defaultdict(list)
```
Comment (WillAyd) [inline pandas/core/aggregation.py]: Is this annotation the only thing that differs from the original code?

Comment (charlesdong1991) [reply pandas/core/aggregation.py]: ahh, yeah, basically! 
also de-private several methods that are used in `generic`, and also de-private methods in docstring as well.

File: pandas/core/aggregation.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+
+# TODO: Can't use, because mypy doesn't like us setting __name__
+#   error: "partial[Any]" has no attribute "__name__"
+# the type is:
+#   typing.Sequence[Callable[..., ScalarResult]]
+#     -> typing.Sequence[Callable[..., ScalarResult]]:
+
+
+def _managle_lambda_list(aggfuncs: Sequence[Any]) -> Sequence[Any]:
```
Comment (WillAyd) [inline pandas/core/aggregation.py]: Would be nice to clean up types in a follow up (these are all Callable right?)

**Review by WillAyd — APPROVED** [review]
lgtm @jreback

File: pandas/core/aggregation.py (hunk 2 above)
Comment (charlesdong1991) [reply pandas/core/aggregation.py]: sure! would come up with follow-up PR for types for all methods in this file, they seem all could be improved a bit

**Comment by charlesdong1991** [comment]
@jreback could you take a look? pls lemme know what changes you want to see here? and then I could start finishing up #29116 after this is merged. Thanks!

File: pandas/core/aggregation.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
+from collections import defaultdict
+from functools import partial
+from typing import Any, DefaultDict, Sequence
+
+from pandas.core.dtypes.common import is_dict_like, is_list_like
+
+import pandas.core.common as com
+from pandas.core.indexes.api import Index
+
```
Comment (jreback) [inline pandas/core/aggregation.py]: can you add a doc-string to the module, describing the purpose of it 

e.g. functions to handle multiple named and lambda kwarg aggregations in groupby and dataframe/series.agg

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, some added typing and fixups of the doc-strings would be good. ping on green.

we can do other things in followons (e.g. making actual logic changes).

File: pandas/tests/groupby/aggregate/test_aggregate.py [hunk 4]
```diff
@@ -633,28 +633,28 @@ def test_lambda_named_agg(func):
 
 class TestLambdaMangling:
```
Comment (jreback) [inline pandas/tests/groupby/aggregate/test_aggregate.py]: move all of these tests to pandas/tests/test_aggregation.py

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 4 above)
Comment (jreback) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: just the tests where you are moving the code (e.g. the maybe_mangle.....) and such

File: pandas/core/aggregation.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        len(kwargs) > 0
+    )
+
+
+def normalize_keyword_aggregation(kwargs):
+    """
+    Normalize user-provided "named aggregation" kwargs.
+    Transforms from the new ``Mapping[str, NamedAgg]`` style kwargs
+    to the old Dict[str, List[scalar]]].
+    Parameters
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/core/aggregation.py]: thanks! added blank line before each section for all functions here and add types for kwargs and add return annotation

File: pandas/core/aggregation.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        for aggfunc in aggfuncs
+    ]
+    uniquified_aggspec = _make_unique(aggspec_order)
+
+    # get the new indice of columns by comparison
+    col_idx_order = Index(uniquified_aggspec).get_indexer(uniquified_order)
+    return aggspec, columns, col_idx_order
+
+
+def _make_unique(seq):
```
[the remark this answers is not in the document]
Comment (charlesdong1991) [reply pandas/core/aggregation.py]: sure, changed!

File: pandas/core/aggregation.py (hunk 3 above)
Comment (charlesdong1991) [reply pandas/core/aggregation.py]: ahh, nice idea, added!

File: pandas/tests/groupby/aggregate/test_aggregate.py (hunk 4 above)
Comment (charlesdong1991) [reply pandas/tests/groupby/aggregate/test_aggregate.py]: yeah, I was thinking it as a follow-up, but maybe good to have it in a go here! I moved all tests related to functions in aggregation to `pandas/tests/test_aggregation.py`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by jreback** [comment]
thanks @charlesdong1991

**Pull request merged**